src/lib/mina_base/zkapp_command.ml

```diff
@@ -1593,71 +1593,410 @@ let dummy =
+        Signed_single
+    | [ Control.(Signature _ | None_given); Control.(Signature _ | None_given) ]
+      ->
+        Signed_pair
+    | _ ->
+        failwith "zkapp_segment_of_controls: Unsupported combination"
+end)
+
 (* Zkapp_command transactions are filtered using this predicate
    - when adding to the transaction pool
    - in incoming blocks
 *)
-let valid_size ~(genesis_constants : Genesis_constants.t) t : unit Or_error.t =
+let valid_size ~(genesis_constants : Genesis_constants.t) (t : t) :
+    unit Or_error.t =
   let events_elements events =
     List.fold events ~init:0 ~f:(fun acc event -> acc + Array.length event)
   in
-  let ( num_proof_updates
-      , num_updates
-      , num_event_elements
-      , num_sequence_event_elements ) =
-    Call_forest.fold t.account_updates ~init:(0, 0, 0, 0)
-      ~f:(fun
-           (num_proof_zkapp_command, num_zkapp_command, evs_size, seq_evs_size)
-           account_update
-         ->
-        let num_proof_zkapp_command' =
-          if Control.(Tag.equal (tag account_update.authorization) Tag.Proof)
-          then num_proof_zkapp_command + 1
-          else num_proof_zkapp_command
-        in
+  let all_updates, num_event_elements, num_sequence_event_elements =
+    Call_forest.fold t.account_updates
+      ~init:([ Account_update.of_fee_payer (fee_payer_account_update t) ], 0, 0)
+      ~f:(fun (acc, num_event_elements, num_sequence_event_elements)
+              (account_update : Account_update.t) ->
         let account_update_evs_elements =
           events_elements account_update.body.events
         in
         let account_update_seq_evs_elements =
           events_elements account_update.body.sequence_events
         in
-        ( num_proof_zkapp_command'
-        , num_zkapp_command + 1
-        , evs_size + account_update_evs_elements
-        , seq_evs_size + account_update_seq_evs_elements ) )
+        ( account_update :: acc
+        , num_event_elements + account_update_evs_elements
+        , num_sequence_event_elements + account_update_seq_evs_elements ) )
+    |> fun (updates, ev, sev) -> (List.rev updates, ev, sev)
+  in
+  let groups =
+    Update_group.group_by_zkapp_command_rev ([] :: [ all_updates ])
+      ( [ ((), ()) ]
+      :: [ ((), ()) :: List.map all_updates ~f:(fun _ -> ((), ())) ] )
   in
-  let max_proof_zkapp_command = genesis_constants.max_proof_zkapp_command in
-  let max_zkapp_command = genesis_constants.max_zkapp_command in
+  let proof_segments, signed_singles, signed_pairs =
+    List.fold ~init:(0, 0, 0) groups
+      ~f:(fun (proof_segments, signed_singles, signed_pairs) { spec; _ } ->
+        match spec with
+        | Proved ->
+            (proof_segments + 1, signed_singles, signed_pairs)
+        | Signed_single ->
+            (proof_segments, signed_singles + 1, signed_pairs)
+        | Signed_pair ->
+            (proof_segments, signed_singles, signed_pairs + 1) )
+  in
+  let proof_cost = 10.26 in
```

Comment:
might be good to include a comment indicating how these constants were derived

Reply:
These constants duplicate the values in `Mina_compille_config`. Can those values be used here?

If so, the explanatory comment should be in that module.

Reply:
Good catch, these were supposed to be from `genesis_constants`. I'll add a comment in `Mina_compile_config`